design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,509 @@
+allowed ('^'), however regex can be complicated to write and is prone to errors
+(escaping full stops etc.).
+
+An empty or nil field can be considered as being "allow all" for that field.
+
+### Behaviour
+
+In order to keep backwards compatibility, and be in-line with the behaviour of
+existing resources such as network policies, the default action will be to
+"allow all" if no policy resource exists. As soon as a resource exists,
+cert-manager will "deny all" unless the CertificateRequest is accepted by the
+existing policy.
+
+The cert-manager CertificateRequest policy controller will watch
+CertificateRequest resources, which when they are created, will evaluate whether
+they pass approval. This is done by checking that a CertificateRequestPolicy
+which approves this request is [bound](#rbac) to the user who appears in the
+user info field of the request. If *any* CertificateRequestPolicy succeeds, the
+CertificateRequest is marked as *approved*. If *no* CertificateRequestPolicy
+succeeds, the CertificateRequest is marked as *denied*.
+
+### RBAC
+
+
+The binding of CertificateRequestPolicy and CertificateRequests is achieved
+using RBAC. This offers two dimensions to how the policy can be applied-
+namespace and user. A policy can be bound to an entire namespace regardless of
+user (system:authenticated), bound to particular users, or a mixture of both.
+
+When the policy controller evaluates whether a particular policy should be
+checked against a CertificateRequest, it will perform a SubjectAccessReview
+against the user present on the CertificateRequest. Since the
+CertificateRequestPolicy can be bound by both namespace and cluster scope, the
+SubjectAccessReview will be requested for both the namespace the
+CertificateRequest resides in, as well as at the cluster scope. The namespace
+scope will be checked first in an effort to improve performance (a large number
+of CertificateRequestPolicy resources are likely scoped by namespace). The
+SubjectAccessReview will be requested for every single CertificateRequestPolicy
+that exists, until and if a CertificateRequestPolicy which is bound approves the
+request. If no CertificateRequestPolicy approves the request, the
+CertificateRequest is marked as denied.
+
+The SubjectAccessReview will request for the attributes:
+
+```yaml
+Group: policy.cert-manager.io
+Verb: use
+Resource: CertificateRequestPolicy
+Name: <Name of the CertificateRequestPolicy being checked>
+Namespace: <Namespace of the CertificateRequest, empty if checking cluster scope>
+```
+
+---
+
+## External Policy Server
+
+To allow for third parties to implement extensions to policy that the core
+functionality does not provide, the CertificateRequestPolicy resource can list
+ExternalPolicyServers. An ExternalPolicyServer resource represents a single
+policy engine server that can receive policy checks, evaluate the request, and
+then respond as to whether the request is approved, not-approved, or in an error
+state.
```

Comment:
If someone wants their own policy engine, why would they not just build against the `approval` API?

Reply:
That's a valid point, why bother with a gRPC API when we can just let people built a controller for the certificate request objects?

Reply:
I asked this question during yesterday's daily standup. Josh explained that using a gRPC API means that we can chain multiple policy engines.

Using the controller approach, coordinating (e.g., ordering) the possibly multiple policy engines is hard (I would even say impossible...)

By the way, do we think there is a need/use-case for chaining policy engines?

cc @munnerz @JoshVanL

Reply:
We are going to be building the policy-approver out of tree for the time being. This will be experimental, and designed to be quickly iterated on. Perhaps there isn't a use case for chaining policies, but since it is out of tree we can reduce the feedback roundtrip on it and find out.

Reply:
Hopefully use case 4, 5 and 6 does give some idea about why chaining policies can be beneficial.